### PR TITLE
CloudWatch: add option to use StatisticsSet for Timer and Distributio…

### PR DESCRIPTION
**Pull request by codemedian**
Hi all,

I took a stab at publishing Timers and DistributionSummaries as StatisticsSet on the cloudwatch2 registry

This should address
* https://github.com/micrometer-metrics/micrometer/issues/1784
* https://github.com/micrometer-metrics/micrometer/issues/457

Let me know what you think, some immediate thoughts would be to merge the Min/Max window into a single class that takes a Comparator for min/max mode

### CODE REVIEW

File: implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchTimer.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import io.micrometer.core.instrument.distribution.TimeWindowMin;
+import io.micrometer.core.instrument.distribution.pause.PauseDetector;
+import io.micrometer.core.instrument.step.StepTimer;
+import io.micrometer.core.instrument.util.TimeUtils;
+
+import java.util.concurrent.TimeUnit;
+
+public class CloudWatchTimer extends StepTimer {
+
+    private final TimeWindowMin min;
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchTimer.java:29]: As well as `CloudWatchDistributionSummary`, this is specific to CloudWatch; how does this resolve #457?

File: implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchDistributionSummary.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     * @param stepMillis                    step in milliseconds
+     * @param supportsAggregablePercentiles whether it supports aggregable percentiles
+     */
+    public CloudWatchDistributionSummary(Id id, Clock clock, DistributionStatisticConfig distributionStatisticConfig, double scale, long stepMillis, boolean supportsAggregablePercentiles) {
+        super(id, clock, distributionStatisticConfig, scale, stepMillis, supportsAggregablePercentiles);
+        this.min = new TimeWindowMin(clock, distributionStatisticConfig);
+    }
+
+    @Override
+    protected void recordNonNegative(double amount) {
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchDistributionSummary.java:42]: I'm not sure I get why this is overridden.

File: implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchMeterRegistry.java [hunk 3]
```diff
@@ -167,25 +194,54 @@ void sendMetricData(List<MetricDatum> metricData) throws InterruptedException {
...
+                            .build();
+
+                    metrics.add(metricDatum(timer.getId(), toStandardUnit(getBaseTimeUnit().name()), statisticSet));
+                }
+            } else {
+                metrics.add(metricDatum(timer.getId(), "sum", getBaseTimeUnit().name(), timer.totalTime(getBaseTimeUnit())));
+                metrics.add(metricDatum(timer.getId(), "count", StandardUnit.COUNT, count));
+                if (count > 0) {
+                    metrics.add(metricDatum(timer.getId(), "avg", getBaseTimeUnit().name(), timer.mean(getBaseTimeUnit())));
+                    metrics.add(metricDatum(timer.getId(), "max", getBaseTimeUnit().name(), timer.max(getBaseTimeUnit())));
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchMeterRegistry.java:216]: If we publish `min` in one case, should not we publish it in both?

**Review by jonatan-ivanov — CHANGES_REQUESTED** [review]

File: implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchDistributionSummary.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ */
+package io.micrometer.cloudwatch2;
+
+import io.micrometer.core.instrument.Clock;
+import io.micrometer.core.instrument.distribution.DistributionStatisticConfig;
+import io.micrometer.core.instrument.distribution.TimeWindowMin;
+import io.micrometer.core.instrument.step.StepDistributionSummary;
+
+public class CloudWatchDistributionSummary extends StepDistributionSummary {
+    private final TimeWindowMin min;
```
[the remark this answers is not in the document]
Comment (codemedian) [reply implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchDistributionSummary.java:24]: Right, based on [this comment](https://github.com/micrometer-metrics/micrometer/pull/2038#issuecomment-624402612) on the linked PR I wrongly assumed it was only meant for CW and/or specific registries. I guess the `TimeWindowMin` utility is useful for that issue, yet doesn't resolve it, I agree.

The `min` is mandatory in the CloudWatch statisticsSet API. I played around with sending magic numbers, but quickly gave up on that

I hadn't looked at the v1 API tbh, but a quick look at [the API](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/cloudwatch/model/MetricDatum.html#setStatisticValues-com.amazonaws.services.cloudwatch.model.StatisticSet-) suggests it's supported.

File: implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchDistributionSummary.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright 2017 VMware, Inc.
```
[the remark this answers is not in the document]
Comment (codemedian) [reply implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchDistributionSummary.java:2]: What's the convention here? Looks like the header is not consistently added to all files?

I've updated the CW ones that had a header and the `TimeWindowMin` Class with 2021. I could add the notice to the remaining files as well if that's preferred

File: implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchMeterRegistry.java [hunk 6]
```diff
@@ -79,7 +81,11 @@ public CloudWatchMeterRegistry(CloudWatchConfig config, Clock clock,
         super(config, clock);
         this.cloudWatchAsyncClient = cloudWatchAsyncClient;
         this.config = config;
-        
+
+        if (config.useStatisticsSet()) {
+            logger.info("publishing Timer and DistributionSummary as StatisticsSet");
```
[the remark this answers is not in the document]
Comment (codemedian) [reply implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchMeterRegistry.java:86]: done

File: implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchDistributionSummary.java (hunk 2 above)
Comment (codemedian) [reply implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchDistributionSummary.java:42]: I blame it on pressing the wrong key combination :)
deleted

File: implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchMeterRegistry.java (hunk 3 above)
Comment (codemedian) [reply implementations/micrometer-registry-cloudwatch2/src/main/java/io/micrometer/cloudwatch2/CloudWatchMeterRegistry.java:216]: Personally I don't think there's much value publishing `min` as a separate metric. 

I'm totally guessing here, so take that with a pinch of salt, but I can see `min/max/avg` being useful for the [CW Anomaly Detection](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Anomaly_Detection.html) feature, but that wouldn't work with the different aggregates being published as separate metrics anyway.

If you think there's value in adding it to the old way as well I'll add it in

File: micrometer-core/src/main/java/io/micrometer/core/instrument/distribution/TimeWindowMin.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.DoubleSupplier;
+import java.util.function.LongSupplier;
+
+/**
+ * An implementation of a decaying minimum for a distribution based on a configurable ring buffer.
+ *
+ * @author Jon Schneider
+ */
+public class TimeWindowMin {
```
[the remark this answers is not in the document]
Comment (codemedian) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/distribution/TimeWindowMin.java:32]: Done.

In general I prefer composition over inheritance, i.e pass the `BiPredicate` and `initialValue` into the constructor of a generic class. Don't have a strong enough opinion about either approach though so went with the template method

File: implementations/micrometer-registry-cloudwatch2/src/test/java/io/micrometer/cloudwatch2/CloudWatchConfigTest.java [hunk 8]
```diff
@@ -97,4 +97,16 @@ void valid() {
 
         assertThat(config.validate().isValid()).isTrue();
     }
+
+    @Test
+    void useStatisticsSet() {
```
[the remark this answers is not in the document]
Comment (codemedian) [reply implementations/micrometer-registry-cloudwatch2/src/test/java/io/micrometer/cloudwatch2/CloudWatchConfigTest.java:102]: done

**Comment by codemedian** [comment]
Hey @jonatan-ivanov , any update on this?

**Comment by codemedian** [comment]
Hi folks. 

I've had some time and muse again to pick this up and have made some minor changes and an addition

1) I changed from using the config option `useStatisticSet` to `useLegacyPublish` as I believe the old way should be deprecated as non-idiomatic
2) I've set up Timers and Distribution statistics in a way that iff `highResolution` and `!useLegacyPublish` the histogram gauges are not registered and instead the histogram is published to CloudWatch using native cloudwatch histogram push functionality (https://www.google.com/search?client=firefox-b-e&q=cloudwatch+publish+histogram)

Below code snippet is what I set up for it to be exported the way I believe it should be
```Java
public static void main(String[] args) throws InterruptedException {
        ImmutableMap<String, String> config = ImmutableMap.<String, String>builder()
                .put("cloudwatch.namespace", "Test/micrometer")
                .put("cloudwatch.step", Duration.ofSeconds(5).toString())
                .put("cloudwatch.useLegacyPublish", "false")
                .build();

        MeterRegistry meterRegistry = new CloudWatchMeterRegistry(
                config::get,
                Clock.SYSTEM,
                CloudWatchAsyncClient
                        .builder()
                        .region(Region.US_EAST_1)
                        .build()
        );

        meterRegistry.config().meterFilter(new MeterFilter() {
            @Override
            public DistributionStatisticConfig configure(Meter.Id id, DistributionStatisticConfig config) {
                if (id.getName().startsWith("test")) {
                    return DistributionStatisticConfig.builder()
                            .percentilesHistogram(true)
                            .build();

                }
                return config;
            }
        });

        while (true) {
            meterRegistry.timer("test.mytimer")
                    .record(ThreadLocalRandom.current().nextInt(50, 500), TimeUnit.MILLISECONDS);
            Thread.sleep(100);
        }
}
```

And the outcome of this change is that it's now possible to use CloudWatch UI toggles to for min/max/avg as well as the `pXX` numbers as illustrated in below screenshot

Screenshot 2021-08-07 at 14 07 26

Sorry for the change of heart after the initial review, but I believe it's for the better

**Comment by LMnet** [comment]
Any updates on this? It's been a few years...

**Pull request closed without merging**